crates/outbound-networking/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+                .as_str()
+                .try_into()
+                .context("could not convert into a url");
+            match (second_try, first_try.map_err(|e| e.into())) {
+                (Ok(u), _) => Ok(u),
+                // Return an error preferring the error from the first attempt if present
+                (_, Err(e)) | (Err(e), _) => Err(e),
+            }
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum AllowedHosts {
+    All,
+    SpecificHosts(Vec<AllowedHost>),
+}
+
+impl AllowedHosts {
+    pub fn parse<S: AsRef<str>>(hosts: &[S]) -> anyhow::Result<AllowedHosts> {
+        // TODO: do we support this?
+        // if hosts.len() == 1 && hosts[0].as_ref() == "insecure:allow-all" {
+        //     return Ok(Self::All);
+        // }
+        let mut allowed = Vec::with_capacity(hosts.len());
+        for host in hosts {
+            allowed.push(AllowedHost::parse(host)?)
+        }
+        Ok(Self::SpecificHosts(allowed))
+    }
+
+    pub fn allows<U: TryInto<Url>>(&self, url: U) -> bool {
+        match self {
+            AllowedHosts::All => true,
+            AllowedHosts::SpecificHosts(hosts) => {
+                let Ok(url) = url.try_into() else {
+                    return false;
+                };
+                hosts.iter().any(|h| h.allows(&url))
+            }
+        }
+    }
+}
+
+impl Default for AllowedHosts {
+    fn default() -> Self {
+        Self::SpecificHosts(Vec::new())
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct AllowedHost {
+    scheme: Option<String>,
+    host: String,
+    port: u16,
+}
+
+impl AllowedHost {
+    fn parse<U: AsRef<str>>(url: U) -> anyhow::Result<Self> {
+        let url_str = url.as_ref();
+        let url: anyhow::Result<Url> = url_str
+            .try_into()
+            .with_context(|| format!("could not convert {url_str:?} into a url"));
+        let (url, has_scheme) = match url {
+            Ok(url) if url.has_host() => (url, true),
+            first_try => {
+                // If the url doesn't successfully parse try again with an added scheme.
+                // This resolves the ambiguity between 'spin.fermyon.com:80' and 'unix:80'.
+                // Technically according to the spec a valid url *must* contain a scheme. However,
+                // we allow url-like strings without schemes, and we interpret the first part as the host.
+                let second_try = format!("scheme://{url_str}")
+                    .as_str()
+                    .try_into()
+                    .context("could not convert into a url");
+                match (second_try, first_try) {
+                    (Ok(u), _) => (u, false),
+                    // Return an error preferring the error from the first attempt if present
+                    (_, Err(e)) | (Err(e), _) => return Err(e),
+                }
+            }
+        };
+        let host = url.host_str().context("the url has no host")?.to_owned();
```

Comment:
Errors referring to "the URL" may be surprising to the user who thought they were entering a host.